cpp/src/arrow/util/int_util.h

```diff
@@ -99,20 +98,24 @@ Status TransposeInts(const DataType& src_type, const DataType& dest_type,
 /// indices must be nonnegative and strictly less than the passed upper
 /// limit (which is usually the length of an array that is being indexed-into).
 ARROW_EXPORT
-Status CheckIndexBounds(const ArrayData& indices, uint64_t upper_limit);
+Status CheckIndexBounds(const ArraySpan& values, uint64_t upper_limit);
 
 /// \brief Boundscheck integer values to determine if they are all between the
 /// passed upper and lower limits (inclusive). Upper and lower bounds must be
 /// the same type as the data and are not currently casted.
 ARROW_EXPORT
-Status CheckIntegersInRange(const Datum& datum, const Scalar& bound_lower,
+Status CheckIntegersInRange(const ArraySpan& values, const Scalar& bound_lower,
```

Comment:
There's plenty of other places where we should nix the use of Datum because of its bloat. We might consider promoting `ExecValue` to the Datum level (like `ValueSpan` or something)